### PR TITLE
[Diagnostics] Cleanup logic related to argument mismatch 

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
Just a minor clean up, since the logic involving arguments e parameter has already been ported, we can remove some of these FIXME's \o/ 
Also updating the documentation now that `FailureDiagnostics::diagnoseParameterErrors` it's gone.

cc @xedin @hborla

### CODE REVIEW

File: lib/Sema/ConstraintSystem.cpp [hunk 1]
```diff
@@ -3741,21 +3741,13 @@ ConstraintSystem::getFunctionArgApplyInfo(ConstraintLocator *locator) {
     choice = overload->choice;
     rawFnType = overload->openedType;
   } else {
+    assert(isa<CallExpr>(anchor) && "Should always be a CallExpr");
```
Comment (theblixguy) [inline lib/Sema/ConstraintSystem.cpp]: I don’t think this is really necessary as the `cast` below will assert it.

Comment (LucianoPAlmeida) [reply lib/Sema/ConstraintSystem.cpp]: Fixed, Thank you :)

**Review by xedin — APPROVED** [review]
Great, thank you!

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Pull request merged**